src/passes/OptimizeInstructions.cpp

```diff
@@ -907,6 +907,44 @@ struct OptimizeInstructions
       }
     }
 
+    if (curr->op == ExtendUInt32 || curr->op == ExtendSInt32) {
+      if (auto* load = curr->value->dynCast<Load>()) {
+        // i64.extend_i32_s(i32.load(_8|_16)(_u|_s)(x))  =>
+        //    i64.load(_8|_16|_32)(_u|_s)(x)
+        //
+        // i64.extend_i32_u(i32.load(_8|_16)(_u|_s)(x))  =>
+        //    i64.load(_8|_16|_32)(_u|_s)(x)
+        //
+        // but we can't do this in following cases:
+        //
+        //    i64.extend_i32_u(i32.load8_s(x))
+        //    i64.extend_i32_u(i32.load16_s(x))
+        //
+        // this mixed sign/zero extensions can't represent in single
+        // signed or unsigned 64-bit load operation. For example if `load8_s(x)`
+        // return i8(-1) (0xFF) than sign extended result will be
+        // i32(-1) (0xFFFFFFFF) and with zero extension to i64 we got
+        // finally 0x00000000FFFFFFFF. However with `i64.load8_s` in this
+        // situation we got `i64(-1)` (all ones) and with `i64.load8_u` it
+        // will be 0x00000000000000FF.
+        //
+        // Another limitation is atomics which only have unsigned loads.
+        // So we also avoid this only case:
+        //
+        //   i64.extend_i32_s(i32.atomic.load(x))
+        if (!(curr->op == ExtendUInt32 && load->signed_ && load->bytes <= 2) &&
+            !(load->isAtomic && load->bytes == 4 && curr->op == ExtendSInt32)) {
```

Comment:
This duplicates the logic lower down, which makes it a little hard to read. Note how the comment explaining the 4 is after the first place the 4 appears, now.

How about, instead, doing something like this?
```
bool willBeSigned = ...
if (!(willBeSigned && atomic)) {
  if (willBeSigned) {
    signed = true;
  }
  .. do the opt..
}
```

Reply:
It quite hard to exclude mixed signed / unsigned case in this form. Could you provide full example for `willBeSigned`?

Reply:
By `willBeSigned` I just mean the case where we change it to be signed. That is, this code:
```cpp
          if (load->bytes == 4) {
            // Special case for i32.load. In this case signedness depends on
            // extend operation.
            load->signed_ = curr->op == ExtendSInt32;
          }
```
That is the only case in which we turn it to be signed, and the dangerous one if atomics are involved.

Reply:
Here example still not cover special cases like `i64.extend_i32_u(i32.load8_s(x))`:
```c++
bool willBeSigned = curr->op == ExtendSInt32 && load->bytes == 4;
if (!(willBeSigned && load->isAtomic)) {
  if (willBeSigned) {
      load->signed_ = true;
  }
load->type = Type::i64;
return replaceCurrent(load);
```

Reply:
I'm not sure this will be much more clear. Hmm:
```c++
bool willBeSigned = curr->op == ExtendSInt32 && load->bytes == 4;
if (!(willBeSigned && load->isAtomic) &&
    !(curr->op == ExtendUInt32 && load->signed_ && load->bytes <= 2)) {
  if (willBeSigned) {
    load->signed_ = true;
  }
  ...
}
```